subsys/mpsl/cx/1wire/mpsl_cx_1wire.c

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+/**
+ * @file
+ *   This file implements a generic 1-wire coexistence interface.
+ */
+
+#if defined(CONFIG_MPSL_CX_PIN_FORWARDER)
+#include <string.h>
+#include <soc_secure.h>
+#else
+#include <mpsl_cx_abstract_interface.h>
+#endif
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/gpio.h>
+
+#include "hal/nrf_gpio.h"
+#include <nrfx_gpiote.h>
+
+#if DT_NODE_EXISTS(DT_NODELABEL(nrf_radio_coex))
+#define CX_NODE DT_NODELABEL(nrf_radio_coex)
+#else
+#define CX_NODE DT_INVALID_NODE
+#error No enabled coex nodes registered in DTS.
+#endif
+
+#if !(DT_NODE_HAS_COMPAT(CX_NODE, generic_radio_coex_one_wire))
+#error Selected coex node is not compatible with generic-radio-coex-one-wire.
+#endif
+
+#if !defined(CONFIG_MPSL_CX_PIN_FORWARDER)
+
+#define GRANT_PIN_PORT_NO  DT_PROP(DT_GPIO_CTLR(CX_NODE, grant_gpios), port)
+#define GRANT_PIN_PIN_NO   DT_GPIO_PIN(CX_NODE, grant_gpios)
+
+static const nrfx_gpiote_t gpiote =
+	NRFX_GPIOTE_INSTANCE(NRF_DT_GPIOTE_INST(CX_NODE, grant_gpios));
+
+static const struct gpio_dt_spec gra_spec = GPIO_DT_SPEC_GET(CX_NODE, grant_gpios);
+
+static mpsl_cx_cb_t callback;
+static struct gpio_callback grant_cb;
+static uint32_t grant_abs_pin;
+
+static bool grant_pin_is_asserted(void)
+{
+	int ret = gpio_pin_get_dt(&gra_spec);
+
+	__ASSERT(ret >= 0, "Error while reading GPIO state.");
+
+	return ret;
+}
+
+static mpsl_cx_op_map_t granted_ops_map_get(void)
+{
+	mpsl_cx_op_map_t granted_ops = MPSL_CX_OP_IDLE_LISTEN;
+
+	if (grant_pin_is_asserted()) {
+		granted_ops |= MPSL_CX_OP_TX | MPSL_CX_OP_RX;
+	}
+
+	return granted_ops;
+}
+
+static int32_t granted_ops_get(mpsl_cx_op_map_t *granted_ops)
+{
+	*granted_ops = granted_ops_map_get();
+	return 0;
+}
+
+static void gpiote_irq_handler(const struct device *gpiob, struct gpio_callback *cb, uint32_t pins)
+{
+	ARG_UNUSED(gpiob);
+	ARG_UNUSED(cb);
+	ARG_UNUSED(pins);
+
+	static mpsl_cx_op_map_t last_notified;
+	mpsl_cx_op_map_t granted_ops;
+
+	if (callback) {
+		granted_ops = granted_ops_map_get();
+
+		if ((last_notified & (MPSL_CX_OP_RX | MPSL_CX_OP_TX))
```

Comment:
I realized now that this part that tracks whether to call the callback or not doesn't really work. I will update to simply call callback() at every interrupt. The protocol implementation can simply read granted_ops to decide how to handle it.